### PR TITLE
many, many improvements

### PR DESCRIPTION
**Pull request by dgreisen-cfpb**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,98 +1,518 @@
...
+    1) Define the actions your application will perform.
+    2) Write an action transform function for each action. 
+       The transform function takes an action and a CouchDB document,
+       and idempotently transforms the document as required by the action.
+    3) Write a validation function for each action.
+       The validation function takes the action,
+       the user who performed the action (the actor),
+       and the old and new doc. It throws an error if
+       the action was invalid or unauthorized.
+    4) Write an asyncronous worker handler for any actions that result in side-effects.
```
Comment (Scotchester) [inline README.md]: If it matters to you, this doesn't produce a normal ordered list in the rendered view.

File: README.md [hunk 2]
```diff
@@ -1,98 +1,518 @@
...
+
+That's pretty easy! 
+We don't even have to worry about the _id.
+CouchDB will create one for us if we don't set it explicitly.
+
+We want to be able to perform four different actions on this data:
+  1. create a new boulder ('b+')
+  2. start rolling the rock up the hill ('bu')
+  3. make the rock slip away and roll back down the hill ('bd')
+  4. bring the boulder to rest at the bottom of the hill ('br')
```
Comment (Scotchester) [inline README.md]: Should the four action names above be wrapped with backticks instead of single quotes?

**Comment by Scotchester** [comment]
That's one heckuva README update! :+1: 

(Note: I didn't/can't review any of the functional code here.)

File: src/middleware.coffee [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+
+module.exports = 
+
+  auth: (conf) ->
+    SYSTEM_USER = conf.COUCHDB.SYSTEM_USER
+    (req, resp, next) ->
+      # look for admin credentials in basic auth, and if valid, login user as admin.
+      credentials = basic_auth(req);
+      if conf.DEV = 'true' or
```
Comment (sephcoster) [inline src/middleware.coffee:11]: Is this supposed to allow regular user credentials or just SYSTEM_USER or nothing? Looking at the syntax it seems to kick you out if you have valid user credentials that are not admin and are not in conf.DEV.

File: src/promise.coffee [hunk 4]
```diff
@@ -103,4 +106,8 @@ Promise.RestClient = (defaults) ->
 
   return Client
 
-module.exports = Promise
\ No newline at end of file
+Promise.setTimeout = Promise.denodeify((delay, args..., callback) ->
```
Comment (sephcoster) [inline src/promise.coffee:109]: To confirm - this is in here for demos / testing? (outside the diff it appears to be pointing to no newline, but this referes to the Promise.setTimeout function)

**Comment by sephcoster** [comment]
Looks good. Appreciated the additional documentation for the helpers. Any licensing considerations that we need to add to the README given new and/or different deps or are those all stored outside the repo?

File: src/middleware.coffee (hunk 3 above)
Comment (dgreisen-cfpb) [inline src/middleware.coffee:11]: that is correct. It is only allowing SYSTEM_USER if you are not in the dev environment. If you want anything else, then you should create your own auth middleware.

File: src/promise.coffee (hunk 4 above)
Comment (dgreisen-cfpb) [inline src/promise.coffee:109]: correct

**Pull request merged**